python/ray/autoscaler/autoscaler.py

```diff
@@ -130,32 +141,27 @@
     # Map of remote paths to local paths, e.g. {"/tmp/data": "/my/local/data"}
     "file_mounts": (dict, OPTIONAL),
 
-    # List of commands that will be run before `setup_commands`. If docker is
-    # enabled, these commands will run outside the container and before docker
-    # is setup.
-    "initialization_commands": (list, OPTIONAL),
-
-    # List of common shell commands to run to setup nodes.
-    "setup_commands": (list, OPTIONAL),
-
-    # Commands that will be run on the head node after common setup.
-    "head_setup_commands": (list, OPTIONAL),
+    # Commands that will be run when the cluster is first created.
+    "node_setup_commands": COMMANDS_SCHEMA,
 
-    # Commands that will be run on worker nodes after common setup.
-    "worker_setup_commands": (list, OPTIONAL),
+    # Commands that will be run whenever a node is booted.
+    "node_restart_commands": COMMANDS_SCHEMA,
 
-    # Command to start ray on the head node. You shouldn't need to modify this.
-    "head_start_ray_commands": (list, OPTIONAL),
-
-    # Command to start ray on worker nodes. You shouldn't need to modify this.
-    "worker_start_ray_commands": (list, OPTIONAL),
+    # Commands that will be run whenever Ray is (re)started.
+    "ray_restart_commands": COMMANDS_SCHEMA,
```

Comment:
There seems to be a lot of heterogenity in the example Ray configs for how Ray is started. For example, `aws/example-full.yaml` sets a limit on the number of open files; `local/example-full.yaml` disables core dump limit (presumably for debugging). In my own usage I've often used the `--num-cpus` flag to reduce the number of reported CPUs on the head node to ensure it isn't overloaded.

Maybe `ray_head_config` and `ray_worker_config` could capture this, but I think it'd be better for you to implement that as I don't have a clear vision for what this would look like.

Reply:
I also have sometimes put commands like `git pull` in this section so that I can just run `ray up` to restart the cluster with a new version of the code.

Reply:
Number of open files & core dump can and should should be specified elsewhere and overloading `ray up` with `git pull` commands isn't the right API. I would strongly prefer removing these and replacing them with config fields as part of this PR or before merging this because I don't want to make two breaking API changes..

Reply:
Agree makes sense to consolidate all breaking API changes into one release. I'm fine with eliminating the `ray start` command and replacing it with a config, but I'm not that familiar with how `ray start` works so would appreciate some pointers here. Some options I can see:
  1. Auto-construct the `ray start` command-line, and have the user optionally specify additional CLI flags to append (defaults to `""`).
  2. Have the user specify a YAML dict of flags to `ray start`. Some if unspecified are automatically filled in with appropriate values (e.g. `redis-port`). Then construct CLI flag from that.
  3. More restricted version where we decide what CLI flags to expose in the autoscaler.

Reply:
Why change the ray start commands? I think having it explicitly here is the simplest solution. Understanding configurations is not easy for users to do, and you may run into situations that are not supported by the config.